Batch listening does not work for typed Spring messaging messages

**Affects Version(s): 2.2.2.RELEASE**

## Bug report

The following batch listener does not work as expected:

```java
@RabbitListener(queues = "aQueue")
public void receive(List<Message<Foo>> batch) {
  batch.forEach(message -> System.out.println(message.getPayload().toString()));
}
```

Incoming messages fail to get deserialized with the following error message:
```
org.springframework.amqp.rabbit.support.ListenerExecutionFailedException: Listener threw exception
[...]
Caused by: org.springframework.amqp.support.converter.MessageConversionException: Failed to convert Message content
[...]
Caused by: com.fasterxml.jackson.databind.exc.InvalidDefinitionException: Cannot construct instance of `org.springframework.messaging.Message` (no Creators, like default construct, exist): abstract types either need to be mapped to concrete types, have custom deserializer, or contain additional type information
 at [Source: (String)"{"test": 1}"; line: 1, column: 1]
```

## Steps to reproduce

```java
@SpringBootApplication
@EnableRabbit
public class SpringAmqpBatchMessageBug {

  public static void main(String[] args) throws Exception {
    ConfigurableApplicationContext context = SpringApplication.run(SpringAmqpBatchMessageBug.class, args);
    RabbitTemplate rabbitTemplate = context.getBean(RabbitTemplate.class);
    rabbitTemplate.send("exchange", "", msg("{\"test\": 1}"));
    rabbitTemplate.send("exchange", "", msg("{\"test\": 2}"));
    rabbitTemplate.send("exchange", "", msg("{\"test\": 3}"));
    context.getBean(Listener.class).latch.await(60, TimeUnit.SECONDS);
    context.close();
  }

  private static org.springframework.amqp.core.Message msg(String body) throws UnsupportedEncodingException {
    MessageProperties properties = new MessageProperties();
    properties.setContentType("application/json");
    return new org.springframework.amqp.core.Message(body.getBytes(SimpleMessageConverter.DEFAULT_CHARSET), properties);
  }

  @Bean
  public FanoutExchange exchange() { return (FanoutExchange) ExchangeBuilder.fanoutExchange("exchange").build(); }
  @Bean
  public Queue workingBatchQueue() { return new Queue("workingBatchQueue"); }
  @Bean
  public Queue brokenBatchQueue() { return new Queue("brokenBatchQueue"); }
  @Bean
  public Binding workingQueueBinding(Queue workingBatchQueue, FanoutExchange exchange) { return BindingBuilder.bind(workingBatchQueue).to(exchange); }
  @Bean
  public Binding brokenQueueBinding(Queue brokenBatchQueue, FanoutExchange exchange) { return BindingBuilder.bind(brokenBatchQueue).to(exchange); }

  @Bean
  public SimpleRabbitListenerContainerFactory simpleRabbitListenerContainerFactory(
      SimpleRabbitListenerContainerFactoryConfigurer configurer,
      ConnectionFactory connectionFactory) {
    SimpleRabbitListenerContainerFactory factory = new SimpleRabbitListenerContainerFactory();
    configurer.configure(factory, connectionFactory);
    factory.setConnectionFactory(connectionFactory);
    factory.setMessageConverter(new Jackson2JsonMessageConverter());
    factory.setBatchListener(true);
    factory.setConsumerBatchEnabled(true);
    factory.setDeBatchingEnabled(true);
    factory.setBatchSize(3);
    factory.setReceiveTimeout(1000L);
    return factory;
  }

  @Bean
  public Listener listener() {
    return new Listener();
  }

  public static class Listener {

    private final CountDownLatch latch = new CountDownLatch(3);

    @RabbitListener(
        queues = "workingBatchQueue",
        containerFactory = "simpleRabbitListenerContainerFactory"
    )
    public void receiveWorkingBatch(List<Foo> batch) {
      System.out.println("Received batch of size " + batch.size() + " from 'workingBatchQueue'");
      batch.forEach(message -> {
        System.out.println("Message in 'workingBatchQueue' batch: " + message);
        latch.countDown();
      });
    }

    @RabbitListener(
        queues = "brokenBatchQueue",
        containerFactory = "simpleRabbitListenerContainerFactory"
    )
    public void receiveBrokenBatch(List<Message<Foo>> batch) {
      System.out.println("Received batch of size " + batch.size() + " from 'brokenBatchQueue'");
      batch.forEach(message -> {
        System.out.println("Message in 'brokenBatchQueue' batch: " + message.getPayload());
        latch.countDown();
      });
    }

  }

  public static class Foo {
    private int test;

    public int getTest() {
      return test;
    }

    public void setTest(int test) {
      this.test = test;
    }

    @Override
    public String toString() {
      return "test: " + test;
    }
  }

}
```

--- a/spring-rabbit/src/main/java/org/springframework/amqp/rabbit/listener/adapter/MessagingMessageListenerAdapter.java
+++ b/spring-rabbit/src/main/java/org/springframework/amqp/rabbit/listener/adapter/MessagingMessageListenerAdapter.java
@@ -1,5 +1,5 @@
 /*
- * Copyright 2002-2019 the original author or authors.
+ * Copyright 2002-2020 the original author or authors.
  *
  * Licensed under the Apache License, Version 2.0 (the "License");
  * you may not use this file except in compliance with the License.
@@ -381,7 +381,8 @@ public class MessagingMessageListenerAdapter extends AbstractAdaptableMessageLis
 				if (parameterizedType.getRawType().equals(Message.class)) {
 					genericParameterType = ((ParameterizedType) genericParameterType).getActualTypeArguments()[0];
 				}
-				else if (parameterizedType.getRawType().equals(List.class)
+				else if (this.isBatch
+						&& parameterizedType.getRawType().equals(List.class)
 						&& parameterizedType.getActualTypeArguments().length == 1) {
 
 					Type paramType = parameterizedType.getActualTypeArguments()[0];
@@ -392,7 +393,7 @@ public class MessagingMessageListenerAdapter extends AbstractAdaptableMessageLis
 					if (messageHasGeneric) {
 						genericParameterType = ((ParameterizedType) paramType).getActualTypeArguments()[0];
 					}
-					if (this.isBatch) {
+					else {
 						// when decoding batch messages we convert to the List's generic type
 						genericParameterType = paramType;
 					}

--- a/spring-rabbit/src/main/java/org/springframework/amqp/rabbit/listener/adapter/MessagingMessageListenerAdapter.java
+++ b/spring-rabbit/src/main/java/org/springframework/amqp/rabbit/listener/adapter/MessagingMessageListenerAdapter.java
@@ -57,6 +57,7 @@ import com.rabbitmq.client.Channel;
  * @author Stephane Nicoll
  * @author Gary Russell
  * @author Artem Bilan
+ * @author Kai Stapel
  *
  * @since 1.4
  */

--- a/spring-rabbit/src/test/java/org/springframework/amqp/rabbit/annotation/EnableRabbitBatchJsonIntegrationTests.java
+++ b/spring-rabbit/src/test/java/org/springframework/amqp/rabbit/annotation/EnableRabbitBatchJsonIntegrationTests.java
@@ -1,5 +1,5 @@
 /*
- * Copyright 2019 the original author or authors.
+ * Copyright 2019-2020 the original author or authors.
  *
  * Licensed under the Apache License, Version 2.0 (the "License");
  * you may not use this file except in compliance with the License.
@@ -18,12 +18,14 @@ package org.springframework.amqp.rabbit.annotation;
 
 import static org.assertj.core.api.Assertions.assertThat;
 
+import java.io.UnsupportedEncodingException;
 import java.util.List;
 import java.util.concurrent.CountDownLatch;
 import java.util.concurrent.TimeUnit;
 
 import org.junit.jupiter.api.Test;
 
+import org.springframework.amqp.core.MessageProperties;
 import org.springframework.amqp.rabbit.batch.SimpleBatchingStrategy;
 import org.springframework.amqp.rabbit.config.SimpleRabbitListenerContainerFactory;
 import org.springframework.amqp.rabbit.connection.CachingConnectionFactory;
@@ -32,6 +34,7 @@ import org.springframework.amqp.rabbit.core.BatchingRabbitTemplate;
 import org.springframework.amqp.rabbit.junit.RabbitAvailable;
 import org.springframework.amqp.rabbit.junit.RabbitAvailableCondition;
 import org.springframework.amqp.support.converter.Jackson2JsonMessageConverter;
+import org.springframework.amqp.support.converter.SimpleMessageConverter;
 import org.springframework.beans.factory.annotation.Autowired;
 import org.springframework.context.annotation.Bean;
 import org.springframework.context.annotation.Configuration;
@@ -58,21 +61,27 @@ public class EnableRabbitBatchJsonIntegrationTests {
 	private Listener listener;
 
 	@Test
-	public void testSimpleList() throws InterruptedException {
-		this.template.convertAndSend("json.batch.1", new Foo("foo"));
-		this.template.convertAndSend("json.batch.1", new Foo("bar"));
+	public void testSimpleList() throws Exception {
+		this.template.send("json.batch.1", msg("{\"bar\":\"foo\"}"));
+		this.template.send("json.batch.1", msg("{\"bar\":\"bar\"}"));
 		assertThat(this.listener.foosLatch.await(10, TimeUnit.SECONDS)).isTrue();
 		assertThat(this.listener.foos.get(0).getBar()).isEqualTo("foo");
 		assertThat(this.listener.foos.get(1).getBar()).isEqualTo("bar");
 	}
 
 	@Test
-	public void testMessageList() throws InterruptedException {
-		this.template.convertAndSend("json.batch.2", new Foo("foo"));
-		this.template.convertAndSend("json.batch.2", new Foo("bar"));
+	public void testMessageList() throws Exception {
+		this.template.send("json.batch.2", msg("{\"bar\":\"foo\"}"));
+		this.template.send("json.batch.2", msg("{\"bar\":\"bar\"}"));
 		assertThat(this.listener.fooMessagesLatch.await(10, TimeUnit.SECONDS)).isTrue();
 		assertThat(this.listener.fooMessages.get(0).getPayload().getBar()).isEqualTo("foo");
 		assertThat(this.listener.fooMessages.get(1).getPayload().getBar()).isEqualTo("bar");
+	}
+
+	private org.springframework.amqp.core.Message msg(String body) throws UnsupportedEncodingException {
+		MessageProperties properties = new MessageProperties();
+		properties.setContentType("application/json");
+		return new org.springframework.amqp.core.Message(body.getBytes(SimpleMessageConverter.DEFAULT_CHARSET), properties);
 	}
 
 	@Configuration
@@ -84,6 +93,8 @@ public class EnableRabbitBatchJsonIntegrationTests {
 			SimpleRabbitListenerContainerFactory factory = new SimpleRabbitListenerContainerFactory();
 			factory.setConnectionFactory(connectionFactory());
 			factory.setBatchListener(true);
+			factory.setConsumerBatchEnabled(true);
+			factory.setBatchSize(2);
 			factory.setMessageConverter(converter());
 			return factory;
 		}

--- a/spring-rabbit/src/test/java/org/springframework/amqp/rabbit/annotation/EnableRabbitBatchJsonIntegrationTests.java
+++ b/spring-rabbit/src/test/java/org/springframework/amqp/rabbit/annotation/EnableRabbitBatchJsonIntegrationTests.java
@@ -46,6 +46,7 @@ import org.springframework.test.context.junit.jupiter.SpringJUnitConfig;
 
 /**
  * @author Gary Russell
+ * @author Kai Stapel
  * @since 2.2
  *
  */

--- a/spring-rabbit/src/test/java/org/springframework/amqp/rabbit/listener/adapter/MessagingMessageListenerAdapterTests.java
+++ b/spring-rabbit/src/test/java/org/springframework/amqp/rabbit/listener/adapter/MessagingMessageListenerAdapterTests.java
@@ -1,5 +1,5 @@
 /*
- * Copyright 2014-2019 the original author or authors.
+ * Copyright 2014-2020 the original author or authors.
  *
  * Licensed under the Apache License, Version 2.0 (the "License");
  * you may not use this file except in compliance with the License.
@@ -25,8 +25,10 @@ import static org.mockito.Mockito.verify;
 
 import java.lang.reflect.Method;
 import java.util.ArrayList;
+import java.util.Arrays;
 import java.util.LinkedHashMap;
 import java.util.List;
+import java.util.stream.Collectors;
 
 import org.junit.jupiter.api.BeforeEach;
 import org.junit.jupiter.api.Test;
@@ -51,6 +53,7 @@ import com.rabbitmq.client.Channel;
  * @author Stephane Nicoll
  * @author Artem Bilan
  * @author Gary Russell
+ * @author Kai Stapel
  */
 public class MessagingMessageListenerAdapterTests {
 
@@ -212,6 +215,54 @@ public class MessagingMessageListenerAdapterTests {
 		assertThat(this.sample.payload.getClass()).isEqualTo(LinkedHashMap.class);
 	}
 
+	@Test
+	public void batchAmqpMessagesTest() {
+		// given
+		org.springframework.amqp.core.Message message1 = MessageTestUtils.createTextMessage("{ \"foo1\" : \"bar1\" }");
+		message1.getMessageProperties().setContentType("application/json");
+		Channel channel = mock(Channel.class);
+		BatchMessagingMessageListenerAdapter listener = getBatchInstance("withAmqpMessageBatch");
+		listener.setMessageConverter(new Jackson2JsonMessageConverter());
+
+		// when
+		listener.onMessageBatch(Arrays.asList(message1), channel);
+
+		// then
+		assertThat(this.sample.batchPayloads.get(0).getClass()).isEqualTo(String.class);
+	}
+
+	@Test
+	public void batchTypedMessagesTest() {
+		// given
+		org.springframework.amqp.core.Message message1 = MessageTestUtils.createTextMessage("{ \"foo1\" : \"bar1\" }");
+		message1.getMessageProperties().setContentType("application/json");
+		Channel channel = mock(Channel.class);
+		BatchMessagingMessageListenerAdapter listener = getBatchInstance("withTypedMessageBatch");
+		listener.setMessageConverter(new Jackson2JsonMessageConverter());
+
+		// when
+		listener.onMessageBatch(Arrays.asList(message1), channel);
+
+		// then
+		assertThat(this.sample.batchPayloads.get(0).getClass()).isEqualTo(Foo.class);
+	}
+
+	@Test
+	public void batchTypedObjectTest() {
+		// given
+		org.springframework.amqp.core.Message message1 = MessageTestUtils.createTextMessage("{ \"foo1\" : \"bar1\" }");
+		message1.getMessageProperties().setContentType("application/json");
+		Channel channel = mock(Channel.class);
+		BatchMessagingMessageListenerAdapter listener = getBatchInstance("withFooBatch");
+		listener.setMessageConverter(new Jackson2JsonMessageConverter());
+
+		// when
+		listener.onMessageBatch(Arrays.asList(message1), channel);
+
+		// then
+		assertThat(this.sample.batchPayloads.get(0).getClass()).isEqualTo(Foo.class);
+	}
+
 	protected MessagingMessageListenerAdapter getSimpleInstance(String methodName, Class<?>... parameterTypes) {
 		Method m = ReflectionUtils.findMethod(SampleBean.class, methodName, parameterTypes);
 		return createInstance(m, false);
@@ -246,6 +297,17 @@ public class MessagingMessageListenerAdapterTests {
 		return adapter;
 	}
 
+	protected BatchMessagingMessageListenerAdapter getBatchInstance(String methodName) {
+		Method m = ReflectionUtils.findMethod(SampleBean.class, methodName, List.class);
+		return createBatchInstance(m);
+	}
+
+	protected BatchMessagingMessageListenerAdapter createBatchInstance(Method m) {
+		BatchMessagingMessageListenerAdapter adapter = new BatchMessagingMessageListenerAdapter(null, m, false, null, null);
+		adapter.setHandlerAdapter(new HandlerAdapter(factory.createInvocableHandlerMethod(sample, m)));
+		return adapter;
+	}
+
 	private void initializeFactory(DefaultMessageHandlerMethodFactory factory) {
 		factory.setBeanFactory(new StaticListableBeanFactory());
 		factory.afterPropertiesSet();
@@ -254,6 +316,7 @@ public class MessagingMessageListenerAdapterTests {
 	private static class SampleBean {
 
 		private Object payload;
+		private List<Object> batchPayloads;
 
 		SampleBean() {
 		}
@@ -293,6 +356,21 @@ public class MessagingMessageListenerAdapterTests {
 		@SuppressWarnings("unused")
 		public void withNonGenericMessage(@SuppressWarnings("rawtypes") Message message) {
 			this.payload = message.getPayload();
+		}
+
+		@SuppressWarnings("unused")
+		public void withAmqpMessageBatch(List<org.springframework.amqp.core.Message> messageBatch) {
+			this.batchPayloads = messageBatch.stream().map(m -> new String(m.getBody())).collect(Collectors.toList());
+		}
+
+		@SuppressWarnings("unused")
+		public void withTypedMessageBatch(List<Message<Foo>> messageBatch) {
+			this.batchPayloads = messageBatch.stream().map(Message::getPayload).collect(Collectors.toList());
+		}
+
+		@SuppressWarnings("unused")
+		public void withFooBatch(List<Foo> messageBatch) {
+			this.batchPayloads = new ArrayList<>(messageBatch);
 		}
 
 		@SuppressWarnings("unused")

